feat: new virtual module `astro:config` (#13084)
withastro/astro: .changeset/stupid-glasses-switch.md

---
'astro': minor
---

> Adds a new virtual module called `astro:config` as part of the relative [RFC Stage 3](https://github.com/withastro/roadmap/pull/1106).

Reviewer: ```suggestion
Adds a new experimental virtual module `astro:config` that exposes a type-safe subset of your `astro.config.mjs` configuration
```
Can probably combine with the sentence below, and leave the links for the bottom so peple don't immediately click away 😄 ?

> The new virtual module exposes a type-safe sub-set configuration defined in `astro.config.mjs`.
> 
> The virtual module exposes two sub-paths:
> - `astro:config/client`: exposes config information that are safe to expose to the client.
> - `astro:config/server`: exposes additional information that is safe to expose to the server, such as file/dir paths.

Reviewer: ```suggestion
The virtual module exposes two sub-paths for controlled access to your configuration:
- `astro:config/client`: exposes config information that is safe to expose to the client.
- `astro:config/server`: exposes additional information that is safe to expose to the server, such as file/dir paths.
```

> To enable this new virtual module, an experimental flag must be turned on:

Reviewer: ```suggestion
To enable this new virtual module, add the `experimental.serializeManifest` feature flag to your Astro config:
```

```js
// astro.config.mjs
import {defineConfig} from "astro/config"
export default defineConfig({
  experimental: {
    serializeManifest: true
  }
})
```

> Then, you can use the module in any file inside the Astro project

Reviewer: ```suggestion
Then, you can access the module in any file inside your project to import and use values from your Astro config:
```
Just making sure that *any* file includes framework files, too! If not, we should describe which ones are exclusions.

```js
// src/utils.js
import { trailingSlash } from "astro:config/client";

function addForwardSlash(path) {
  if (trailingSlash === "always") {
    return path.endsWith("/") ? path : path + "/"
  } else {
    return path
  }
}

```

> 

Reviewer: ```suggestion

For a complete overview, and to give feedback on this experimental API, see the [Serialized Manifest RFC](https://github.com/withastro/roadmap/blob/feat/serialised-config/proposals/0051-serialized-manifest.md).
```

Noted in adding this standard line, the Stage 3 PR 1106 hasn't yet been added to the description there.
